benchmark/coniql_performance_test.py

```diff
@@ -0,0 +1,366 @@
+        )
+
+        protocols = ["graphql-ws", "graphql-transport-ws"]
+
+        async with websockets.connect(
+            self.endpoint,
+            subprotocols=[protocols[self.ws_protocol - 1]],
+        ) as websocket:
+            await websocket.send(connection_init_message)
+            if self.ws_protocol == 2:
+                await websocket.send(request_message_graphql_transport_ws_protocol)
+            else:
+                await websocket.send(request_message_graphql_ws_protocol)
+
+            msg_count = 0
+            start_time = time.time()
+            async for response in websocket:
+                data = json.loads(response)
+                if data["type"] == "connection_ack":
+                    pass
+                elif data["type"] == "ka":
+                    pass
+                else:
+                    if self.signal.get_start():
+                        handle(data["payload"])
+                        if n_messages is None:
+                            # Do nothing and continue subscription indefinitely
+                            pass
+                        else:
+                            if msg_count > n_messages:
+                                if self.log:
+                                    self.progress_file.close()
+                                break
+                            msg_count = msg_count + 1
+                            if monitor_progress:
+                                ten_percent = round(n_messages / 10, 0)
+                                if msg_count % ten_percent == 0:
+                                    remaining_time = (
+                                        (time.time() - start_time)
+                                        / msg_count
+                                        * (n_messages - msg_count)
+                                    )
+                                    self.progress_file.write(
+                                        f"Collected {msg_count}/{n_messages} samples.\
+                                            Remaining time: {round(remaining_time, 0)}\
+                                            secs \n"
+                                    )
+                                    self.progress_file.flush()
+                    else:
+                        continue
+
+
+class StartStopSignal:
+    def __init__(self):
+        self.start = False
+        self.stop = False
+
+    def signal_start(self):
+        print("-> Starting monitor")
+        self.start = True
+
+    def signal_stop(self):
+        print("-> Stopping monitor")
+        self.stop = True
+
+    def get_start(self):
+        return self.start
+
+    def get_stop(self):
+        return self.stop
+
+
+class PVSubscription:
+    def __init__(self, pv):
+        self.pv = pv
+        self.values = []
+
+    def append(self, value):
+        self.values.append(value)
+
+    def get_values(self):
+        return self.values
+
+
+def cpu_monitor(signal):
```

Comment:
A minor niggle: Every instance of this client program will spawn this same CPU monitor. All but the first are then giving redundant information. The right fix is probably to turn this into a standalone program, but at this point that may not be worth the effort.

Reply:
Attaching creating this to the logging flag should work.

Reply:
I have added a separate argument for the Python script named `--no-cpu-monitor`. If this is provided then the CPU monitor will not be started. I thought that the default behaviour should be that the CPU monitor runs, even if a log file for the progress is not provided and so I added the extra argument.
Within the bash script, the first client specifies a log file so will monitor subscription progress and will monitor the CPU, while subsequent clients will not run a CPU monitor.